Better cast/convert support for Cosmos

ChrisJollyAU opened the pull request:
Improve the support and handling of the cast/converts in Cosmos

Originally started for #34963 but after #34982 was created extended that work to more cases

`VisitUnary` now leaves the `Convert` operation in place instead of stripping it. This has 2 potential benefits:
- Similarity between Cosmos and relational may help any efforts to create a common shared base rather than a lot of duplication of things (e.g. SqlExpressionFactory etc) in Cosmos
- Leaving it in place allows binary expressions to work out what type mapping the Convert needs (if it doesn't have one already)
- This is especially the case with string concatenation to an int or something like that. the `int` value will have a Convert to type of `Object` - aka we know we need a convert but we don't yet know what to convert it to. The type mapping handling of the binary expression will work it out and apply to all sides

In the QuerySqlGenerator, we effectively no-op most casts except for if we have a convert to string.

This does light up a number of math function tests that were failing - all of them due to #17246 (client evaluation exception)
Also some tests for #20677 are now passing (not all of them related to that issue)
The Cosmos tests for #34729 are now passing as well

This fixes #34982

ChrisJollyAU committed:
Handle string concatenation when there is a convert operation on either side

ChrisJollyAU committed:
Handle cast/convert in Cosmos better

ChrisJollyAU committed:
Further improvements

ChrisJollyAU committed:
merge

ChrisJollyAU committed:
Update tests

ChrisJollyAU commented:
@roji Was just working on this to update and resync everything especially because of all your reorganization of the tests and have a question.

In the original NorthwindFunctions you have the test `Where_mathf_floor` which fails but the new test `Floor_float` passes. There is in fact a difference between the 2 and that is that there is a cast to the function parameter.

Can be summarized by this test (based on `Floor_float`)
```
[ConditionalTheory]
[MemberData(nameof(IsAsyncData))]
public virtual Task Floor_float_double(bool async)
    => AssertQuery(
        async,
        ss => ss.Set<BasicTypesEntity>().Where(b => MathF.Floor((float)b.Double) == 8));
```

In the current `main` branch this will fail (not translated) but with this PR it does work.
Is there any test in the new set of tests that does test this sort of cast/convert as the transition to `MathTranslations` seems to have simplified that aspect out?

roji commented:
> To me it might make more sense to do it in the same place as the original i.e. you have one test normal with direct value and the next test with the cast. Then you end up dealing with the same tables and data and if you modify something for one, you have all the similar tests close by to modify as well

The thing is that one test here (the missing one) tests only the downcast, without it mattering at all which function (or even type) is involved, whereas the other one tests translation for a specific function on a specific type (MathF.Floor), no? Unless you think there is (or could be) a specific issue with the translation of MathF.Floor around the cast?

Also, note that all the translation tests already use the same tables and data. The idea is to try to be in a place where data can be added to accomodate new tests, without that breaking existing tests (it'll likely not be 100% successful, but that's the idea).

ChrisJollyAU commented:
> Unless you think there is (or could be) a specific issue with the translation of MathF.Floor around the cast?

Pretty sure its just the cast itself and doesn't matter what function its used in.

But this is more of a Cosmos specific issue. Relational already has this sorted out since practically v1

Also, there are a bunch of other tests that seem to test this intrinsically and pass with this fix

e.g. in NorthwindFunctions `Where_functions_nested` you have a `Length` (int) within a `Math.Pow` (double)

So a specific test might not be needed

ChrisJollyAU commented:
@roji Can see how this goes now. Should be updated and back in sync

ChrisJollyAU committed:
fix some missed tests

ChrisJollyAU commented:
@roji Any update on this?

ChrisJollyAU commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -473,6 +473,12 @@ protected override ShapedQueryExpression TranslateCast(ShapedQueryExpression sou
+        {
+            item = unaryExpression.Operand;
+        }
         // Simplify x.Array.Contains[1] => ARRAY_CONTAINS(x.Array, 1) insert of IN+subquery
```
Looks fine

ChrisJollyAU replied:
Has to be `if (item is UnaryExpression { NodeType: ExpressionType.Convert or ExpressionType.ConvertChecked } unaryExpression && unaryExpression.Type == typeof(object))` otherwise compile error

ChrisJollyAU commented on src/EFCore.Cosmos/Query/Internal/CosmosSqlTranslatingExpressionVisitor.cs:
```diff
@@ -185,10 +191,6 @@ when TryRewriteEntityEquality(
                     out var result):
                 return result;
-
-            case { Method: var method } when method == ConcatMethodInfo:
-                return QueryCompilationContext.NotTranslatedExpression;
-
             default:
                 var uncheckedNodeTypeVariant = binaryExpression.NodeType switch
```
Looks fine

ChrisJollyAU committed:
merge

ChrisJollyAU committed:
conflicts

ChrisJollyAU committed:
fix sql naselines

ChrisJollyAU committed:
Add test for case-insensitive parameter uniquification

Add an override for Two_parameters_with_same_case_insensitive_name_get_uniquified in NorthwindWhereQueryCosmosTest.cs. The test verifies that SQL parameter names are correctly uniquified when they differ only by case.

ChrisJollyAU committed:
dont need that test anymore

ChrisJollyAU committed:
and this line

AndriySvyryd commented on src/EFCore.Cosmos/Query/Internal/CosmosSqlTranslatingExpressionVisitor.cs:
```diff
@@ -832,34 +834,48 @@ protected override Expression VisitUnary(UnaryExpression unaryExpression)
+                {
+                    sqlOperand = sqlExpressionFactory.ApplyDefaultTypeMapping(sqlOperand);
+
+                    return sqlExpressionFactory.Convert(sqlOperand!, unaryExpression.Type);
```
> - PrimitiveCollectionsQueryCosmosTest.Parameter_collection_in_subquery_and_Convert_as_compiled_query

We can preserve the arguably better exception message by bailing here if `sqlOperand?.TypeMapping is null`

ChrisJollyAU committed:
fix issues

AndriySvyryd reviewed (approved):
Thanks for your contribution!
